File: tauri-agent-demo/src/shared/api/internal/sessions.ts
import type {
  ChatSession,
  ChatSessionCreate,
  ChatSessionUpdate,
  LLMCall,
  Message,
  SessionToolStats,
} from '../../../types';
import { API_BASE_URL } from './base';
import type { AgentStepWithMessage } from './streaming';

export async function getSessions(): Promise<ChatSession[]> {
    const response = await fetch(`${API_BASE_URL}/sessions`);
    if (!response.ok) throw new Error('Failed to fetch sessions');
    return response.json();
}

export async function getSession(
    sessionId: string,
    options?: { includeCount?: boolean }
): Promise<ChatSession> {
    const params = new URLSearchParams();
    if (options?.includeCount === false) {
        params.set('include_count', 'false');
    }
    const query = params.toString();
    const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}${query ? `?${query}` : ''}`);
    if (!response.ok) throw new Error('Failed to fetch session');
    return response.json();
}

export async function createSession(session: ChatSessionCreate): Promise<ChatSession> {
    const response = await fetch(`${API_BASE_URL}/sessions`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(session),
    });
    if (!response.ok) throw new Error('Failed to create session');
    return response.json();
}

export async function updateSession(sessionId: string, update: ChatSessionUpdate): Promise<ChatSession> {
    const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(update),
    });
    if (!response.ok) throw new Error('Failed to update session');
    return response.json();
}

export async function deleteSession(sessionId: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}`, {
        method: 'DELETE',
    });
    if (!response.ok) throw new Error('Failed to delete session');
}

export async function copySession(sessionId: string): Promise<ChatSession> {
    const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}/copy`, {
        method: 'POST',
    });
    if (!response.ok) throw new Error('Failed to copy session');
    return response.json();
}

export async function getSessionMessages(
    sessionId: string,
    options?: { limit?: number; beforeId?: number }
): Promise<Message[]> {
    const params = new URLSearchParams();
    if (options?.limit) {
        params.set('limit', String(options.limit));
    }
    if (options?.beforeId) {
        params.set('before_id', String(options.beforeId));
    }
    const query = params.toString();
    const url = query
        ? `${API_BASE_URL}/sessions/${sessionId}/messages?${query}`
        : `${API_BASE_URL}/sessions/${sessionId}/messages`;
    const response = await fetch(url);
    if (!response.ok) throw new Error('Failed to fetch messages');
    return response.json();
}

export async function getSessionLLMCalls(sessionId: string): Promise<LLMCall[]> {
    const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}/llm_calls`);
    if (!response.ok) throw new Error('Failed to fetch LLM calls');
    return response.json();
}

export async function getSessionToolStats(sessionId: string): Promise<SessionToolStats> {
    const response = await fetch(`${API_BASE_URL}/sessions/${sessionId}/tool_stats`);
    if (!response.ok) throw new Error('Failed to fetch tool stats');
    return response.json();
}

export async function getSessionAgentSteps(
    sessionId: string,
    messageIds?: number[]
): Promise<AgentStepWithMessage[]> {
    const params = new URLSearchParams();
    if (messageIds && messageIds.length > 0) {
        params.set('message_ids', messageIds.join(','));
    }
    const query = params.toString();
    const url = query
        ? `${API_BASE_URL}/sessions/${sessionId}/agent_steps?${query}`
        : `${API_BASE_URL}/sessions/${sessionId}/agent_steps`;
    const response = await fetch(url);
    if (!response.ok) throw new Error('Failed to fetch agent steps');
    return response.json();
}

File: tauri-agent-demo/python-backend/server/routers/sessions.py
from typing import List

from fastapi import APIRouter

from models import ChatMessage, ChatSession
from server.services import base_service
from server.services import session_service


router = APIRouter(tags=["sessions"])
router.add_api_route("/sessions", session_service.get_sessions, methods=["GET"], response_model=List[ChatSession])
router.add_api_route("/sessions/{session_id}", session_service.get_session, methods=["GET"], response_model=ChatSession)
router.add_api_route("/sessions", session_service.create_session, methods=["POST"], response_model=ChatSession)
router.add_api_route("/sessions/{session_id}", session_service.update_session, methods=["PUT"], response_model=ChatSession)
router.add_api_route("/sessions/{session_id}/copy", session_service.copy_session, methods=["POST"], response_model=ChatSession)
router.add_api_route("/sessions/{session_id}", session_service.delete_session, methods=["DELETE"])
router.add_api_route("/sessions/{session_id}/messages", session_service.get_session_messages, methods=["GET"], response_model=List[ChatMessage])
router.add_api_route("/sessions/{session_id}/llm_calls", session_service.get_session_llm_calls, methods=["GET"])
router.add_api_route("/sessions/{session_id}/tool_stats", session_service.get_session_tool_stats, methods=["GET"])
router.add_api_route("/sessions/{session_id}/agent_steps", session_service.get_session_agent_steps, methods=["GET"])
router.add_api_route("/sessions/{session_id}/rollback", session_service.rollback_session, methods=["POST"])
router.add_api_route("/attachments/{attachment_id}", base_service.get_attachment, methods=["GET"])
